### PR TITLE
add static library

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
this can be useful if others want to embed snmalloc into their programs. 

For example: https://github.com/antirez/redis/pull/7262

### CODE REVIEW

**Comment by mjp41** [comment]
Could you make the CI build the static library on as many platforms as makes sense? Otherwise looks great.  Does this make sense as an option for Windows even?

**Comment by SchrodingerZhu** [comment]
mimalloc compiles both dynamic and static library on all platforms by default. Maybe I can just set this option to `On` by default?

**Comment by davidchisnall** [comment]
I am curious as to the intended use case for this?  If you are deeply embedding snmalloc, it is a header-only project that you can just include without needing a standard library.  The shim library is built to be `LD_PRELOAD`ed, but you can't do that with a static library.  The only use case I can see for this is if you are statically linking and replacing the system `malloc`, but that seems a somewhat dangerous thing to encourage.

**Comment by SchrodingerZhu** [comment]
@davidchisnall unlike jemalloc and mimalloc, snmalloc do not have a header file define the drop-in replacement like: `sn_malloc`, `sn_free`, maybe providing that will be a better solution?

**Comment by SchrodingerZhu** [comment]
I think we can use `SNMALLOC_NAME_MANGLE` to provide a similar approach like `jemalloc`, but still, this requires building static library.

**Comment by davidchisnall** [comment]
In other projects where I am providing a new malloc API, I define `SNMALLOC_NAME_MANGLE` and then `#include` the override files in a .cc file in my project.  That said, I am also using a custom PAL for these projects, so I would not be able to use a static library.  I don't object to providing one, I'm just curious as to how it would be used.

**Comment by SchrodingerZhu** [comment]
I see, may 
```c++
 c++ -Isrc -c src/override/malloc.cc -mcx16 -std=c++17 -D"SNMALLOC_NAME_MANGLE(x)=sn_##x" -o override.o
```
is good enough to use.

**Comment by mjp41** [comment]
So perhaps have a Cmake option:
   `SNMALLOC_STATIC_LIBRARY_PREFIX`
The default is "sn_", and always build the static library on all platforms.

**Comment by SchrodingerZhu** [comment]
BTW, is it better to output `libsnmallocshim.a` on unix platforms, and `libsnmallocshim-static.a` on Win32 platforms?

see https://github.com/microsoft/mimalloc/blob/master/CMakeLists.txt#L246

**Comment by mjp41** [comment]
I am happy with everything being called `-static`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -230,6 +231,11 @@ if(NOT DEFINED SNMALLOC_ONLY_HEADER_LIBRARY)
 
   if(NOT WIN32)
     set(SHARED_FILES src/override/new.cc src/override/malloc.cc)
+    if (SNMALLOC_STATIC_LIBRARY)
+      add_shim(snmallocshim-static STATIC ${SHARED_FILES})
```
Comment (mjp41) [inline CMakeLists.txt:235]: Should the static lib be compiling `new.cc`  this is overriding the `new` and `delete` operations globablly without namespace?

File: CMakeLists.txt [hunk 2]
```diff
@@ -230,6 +231,11 @@ if(NOT DEFINED SNMALLOC_ONLY_HEADER_LIBRARY)
 
   if(NOT WIN32)
     set(SHARED_FILES src/override/new.cc src/override/malloc.cc)
+    if (SNMALLOC_STATIC_LIBRARY)
```
Comment (mjp41) [inline CMakeLists.txt:234]: Can you lift the static lib outside the `NOT WIN32`

**Pull request merged**